text/0000-let-expression.md

```diff
@@ -0,0 +1,917 @@
+instead of traditional `and`, `or` remind you the short circuit nature of them and help
+understanding them better.*
+
+It also available many super powers for us that can
+help decreasing rightward drift without adding to implementation and understanding complexity, and
+actually decreasing it by removing `let-else` and preventing from future similar constructs.
+
+### New constructs
+
+```rust
+// reuse if let body
+if let Some(x) = a || let Some(x) = b || Ok(x) = c {
+    // body with many lines of code
+} else {
+    // else with many lines of code
+}
+
+// today alternative without code duplication:
+if let ((Some(x), _, _) | (_, Some(x), _) | (_, _, Ok(x)) = (a, b, c) {
+
+// assignment with default
+(let Foo(x) = a)
+|| (let Bar(x) = b)
+|| (let x = default);
+
+// today alternative:
+let x = if let Foo(x) = a {
+    x
+} else if let Bar(x) = b {
+    x
+} else {
+    default
+};
+
+// simple let expression
+assert!((let Some(x) = a) && (let Some(y) = b(x)) && x == y);
+
+// today alternative with if let chains:
+assert!(matches!(a, Some(x) if let Some(y) = b(x) && x == y));
+assert!(if let Some(x) = a && let Some(y) = b(x) && x == y { true } else { false });
+```
+
+### Practical usage of this features
+
+People find let expressions theoretical at first look, and think only traditional cases (if-let-chain and let-else)
+can become useful. In this section there are some usages for new features of this RFC in real codes.
+
+This is an example from rust-clippy repository:
+```rust
+for w in block.stmts.windows(2) {
+    if_chain! {
+        if let StmtKind::Semi(first) = w[0].kind;
+        if let StmtKind::Semi(second) = w[1].kind;
+        if !differing_macro_contexts(first.span, second.span);
+        if let ExprKind::Assign(lhs0, rhs0, _) = first.kind;
+        if let ExprKind::Assign(lhs1, rhs1, _) = second.kind;
+        if eq_expr_value(cx, lhs0, rhs1);
+        if eq_expr_value(cx, lhs1, rhs0);
+        then {
+            // 30 lines of code with massive rightward drift
+        }
+    }
+}
+```
+
+Which by a generalized let-else can become:
+```rust
+for w in block.stmts.windows(2) {
+    let StmtKind::Semi(first) = w[0].kind
+    && let StmtKind::Semi(second) = w[1].kind
+    && !differing_macro_contexts(first.span, second.span)
+    && let ExprKind::Assign(lhs0, rhs0, _) = first.kind
+    && let ExprKind::Assign(lhs1, rhs1, _) = second.kind
+    && eq_expr_value(cx, lhs0, rhs1)
+    && eq_expr_value(cx, lhs1, rhs0)
+    || continue;
+    // 30 lines of code with two less tab
+}
+```
+Every `if let` or `if_chain!` that fill body of a loop or function can refactored in this way. You can easily find
+dozens of them just in rust-clippy. Note that if-let-chain alone can't solve this problem, because you have access to
+variables of if-let inside of block, but you have access to variables of a binding statement under it, thus you can
+save one indentation.
```

Comment:
That is not a strong argument to me. I don't see the value in saving just one indentation. It might just make the code look more confusing.

Reply:
Saving one indentation is the propose of let-else and I assume let-else has value. And this pattern can also become useful in reusing else block in consecutive let-else (I added some text about this). Similar can be added to current let-else RFC in syntax like `let pat = expr, let pat = expr, let pat = expr else { }`. Do you find this syntax (and let-else in general) also confusing?

Reply:
I find let-else to be a specific contained construct, simplifying a very common code pattern (match this pattern or bail out if it doesn't match).

Similarly, I find if-let chaining to be a specific contained construct, matching several things in a row without nesting, simplifying a common pattern.

This, on the other hand, feels like the wrong side of the tradeoffs between expressiveness and readability, and I think it's a rare example of language orthogonality taken too far. Normally, I'm very much in favor of orthogonality, and of general constructs that cover many different cases with the same mechanism. However, I don't think let matching should be a general boolean construct like this. I think the kind of code enabled by this that *isn't* enabled by let-else or if-let chaining is not code we *want* to enable. Every example of it in this RFC seems less readable than the equivalent way of expressing it without this construct.